File: Problems/Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> state;
        int start = 0;
        int res = 0;

        for (int end = 0; end < s.length(); ++end) {
            ++state[s[end]];

            while (state[s[end]] > 1) {
                --state[s[start]];
                ++start;
            }
            res = max(res, end - start + 1);
        }
        return res;
    }
};

File: Problems/Longest Substring Without Repeating Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        int left = 0, maxLen = 0;

        for (int right = 0; right < s.length(); right++) {
            char c = s.charAt(right);
            if (map.containsKey(c) && map.get(c) >= left) {
                left = map.get(c) + 1; // move left past duplicate
            }
            map.put(c, right); // record last index
            maxLen = Math.max(maxLen, right - left + 1); // update answer
        }
        return maxLen;
    }
}
